### ISSUE TITLE
Install issue

### ISSUE THREAD

**Issue by amoradkh** [bohundan/treadscan#1]
I faced by error during the package installation :
Building wheel for treadscan (pyproject.toml): finished with status 'error'
Failed to build treadscan
 ERROR: Command errored out with exit status 1:
   command: 'D:\Amin Doc\python\pythonProject\venv\Scripts\python.exe' 'D:\Amin Doc\python\pythonProject\venv\lib\site-packages\pip\_vendor\pep517\in_process\_in_process.py' build_wheel 'C:\Users\Zahra\AppData\Local\Temp\tmptqzaer36'
       cwd: C:\Users\Zahra\AppData\Local\Temp\pip-install-s_sat0gm\treadscan_e1f0a83b08ef4ab494b8c89a6fca1898
  Complete output (8 lines):
  running bdist_wheel
  running build
  running build_py
  warning: build_py: byte-compiling is disabled, skipping.
running build_ext
  building 'treadscan.cwrap' extension
  error: Microsoft Visual C++ 14.0 or greater is required. Get it with "Microsoft C++ Build Tools": https://visualstudio.microsoft.com/visual-cpp-build-tools/
  ----------------------------------------
  ERROR: Failed building wheel for treadscan
ERROR: Could not build wheels for treadscan, which is required to install pyproject.toml-based projects

**Comment by bohundan** [comment]
Have you tried solutions suggested here: https://stackoverflow.com/questions/44951456/pip-error-microsoft-visual-c-14-0-is-required?

Btw I don't know if the Cython extension actually compiles on windows, seeing as i set this flag for the compiler `os.environ['CFLAGS'] = '-shared -pthread -fPIC -fwrapv -O3 -fno-strict-aliasing -fopenmp -I "/usr/include/python3.8"'` in `setup.py`. The `/usr/include/python3.8` hardcoded like that is really bad practice and I'm not so sure it works if your system doesn't have that python version in that specific path (it's probably not going to be under /usr/ on windows).

I'm not an expert on the whole python packaging ecosystem and was happy enough that I got it to compile correctly on my machine.

**Comment by amoradkh** [comment]
Thanks Bohundan,
I worked on windows 10 and got error again(after installing new version of
MSBuild ):
[image: image.png]
All the best,
Amin

**Comment by bohundan** [comment]
It seems like your attached image didn't upload. Anyways I tested some things and the flags being set in `setup.py` are redundant, I'll remove them and create release 1.0.5 tomorrow. It could be messing with the compiler, but I don't know what error you got this time.

**Comment by amoradkh** [comment]
Thanks Bohundan for your attention. I uploaded the error image again.
All the best

**Comment by bohundan** [comment]
I still don't see it :smile:. Couldn't you copy paste the error again?

**Comment by amoradkh** [comment]
here it is:
Building wheels for collected packages: treadscan
  Building wheel for treadscan (pyproject.toml) ... error
  error: subprocess-exited-with-error

  × Building wheel for treadscan (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [14 lines of output]
      running bdist_wheel
      running build
      [end of output]

  note: This error originates from a subprocess, and is likely not a
problem with pip.
  ERROR: Failed building wheel for treadscan
Failed to build treadscan
ERROR: Could not build wheels for treadscan, which is required to install
pyproject.toml-based projects

**Comment by bohundan** [comment]
Sorry, but that doesn't give me much info. Nonetheless, I've released version 1.0.5., which doesn't have the (possibly mischievous) compiler flags in setup.py anymore. If you run into any trouble, I think it's not specific to the treadscan package. 

I also just now tested installing the newest treadscan version on my machine and `pip install treadscan` works just fine.

**Comment by SimenHasvi** [comment]
Not sure if this is related, but I had to use python 3.8 for the pip install to work, newer python versions just gave an error.

**Comment by amoradkh** [comment]
Thanks Bohundan!

**Comment by bohundan** [comment]
Actually I just tried on a different system and got the issue which made me add the hacky compiler flags in the first place :frowning_face: . The compiler complains about not having `Python.h`, which comes from some `python3-devel` package? And that's where I'm stuck with this
![Untitled](https://user-images.githubusercontent.com/100153558/197611058-10751dd9-967f-4a26-ae7f-934413376516.png)

I don't know if I have to specify some required library in requirements.txt or how you're supposed to link that correctly.

**Comment by amoradkh** [comment]
I tried on another system and got the message:

PS C:\Users\asus\AppData\Local\Temp\scoped_dir9616_1223877761> pip install
treadscan
Collecting treadscan
  Downloading treadscan-1.0.5.tar.gz (135 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 135.1/135.1 kB 2.0 MB/s eta
0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Collecting torch
  Using cached torch-1.12.1-cp310-cp310-win_amd64.whl (162.2 MB)
Collecting torchvision
  Using cached torchvision-0.13.1-cp310-cp310-win_amd64.whl (1.1 MB)
Collecting opencv-contrib-python
  Using cached opencv_contrib_python-4.6.0.66-cp36-abi3-win_amd64.whl (42.5
MB)
Requirement already satisfied: opencv-python in
c:\users\asus\appdata\local\programs\python\python310\lib\site-packages
(from treadscan) (4.6.0.66)
Requirement already satisfied: numpy in
c:\users\asus\appdata\local\programs\python\python310\lib\site-packages
(from treadscan) (1.23.2)
Collecting improutils
  Using cached improutils-1.1.10-py3-none-any.whl
Requirement already satisfied: matplotlib>=3.4.2 in
c:\users\asus\appdata\local\programs\python\python310\lib\site-packages
(from improutils->treads
can) (3.5.3)
Collecting PyYAML>=5.3.1
  Using cached PyYAML-6.0-cp310-cp310-win_amd64.whl (151 kB)
Requirement already satisfied: Pillow>=8.1.0 in
c:\users\asus\appdata\local\programs\python\python310\lib\site-packages
(from improutils->treadscan)
 (9.2.0)
Collecting Pylon>=0.4.4
  Using cached Pylon-0.4.4-py3-none-any.whl
Requirement already satisfied: wheel in
c:\users\asus\appdata\local\programs\python\python310\lib\site-packages
(from improutils->treadscan) (0.37.1
)
Collecting natsort>=5.3.3
  Using cached natsort-8.2.0-py3-none-any.whl (37 kB)
Collecting pytesseract>=0.3.7
  Using cached pytesseract-0.3.10-py3-none-any.whl (14 kB)
Requirement already satisfied: typing-extensions in
c:\users\asus\appdata\local\programs\python\python310\lib\site-packages
(from torch->treadscan)
(4.3.0)
Requirement already satisfied: requests in
c:\users\asus\appdata\local\programs\python\python310\lib\site-packages
(from torchvision->treadscan) (2.
28.1)
Requirement already satisfied: cycler>=0.10 in
c:\users\asus\appdata\local\programs\python\python310\lib\site-packages
(from matplotlib>=3.4.2->impr
outils->treadscan) (0.11.0)
Requirement already satisfied: fonttools>=4.22.0 in
c:\users\asus\appdata\local\programs\python\python310\lib\site-packages
(from matplotlib>=3.4.2-
>improutils->treadscan) (4.37.1)
Requirement already satisfied: pyparsing>=2.2.1 in
c:\users\asus\appdata\local\programs\python\python310\lib\site-packages
(from matplotlib>=3.4.2->
improutils->treadscan) (3.0.9)
Requirement already satisfied: kiwisolver>=1.0.1 in
c:\users\asus\appdata\local\programs\python\python310\lib\site-packages
(from matplotlib>=3.4.2-
>improutils->treadscan) (1.4.4)
Requirement already satisfied: packaging>=20.0 in
c:\users\asus\appdata\local\programs\python\python310\lib\site-packages
(from matplotlib>=3.4.2->i
mproutils->treadscan) (21.3)
Requirement already satisfied: python-dateutil>=2.7 in
c:\users\asus\appdata\local\programs\python\python310\lib\site-packages
(from matplotlib>=3.4
.2->improutils->treadscan) (2.8.2)
Requirement already satisfied: idna<4,>=2.5 in
c:\users\asus\appdata\local\programs\python\python310\lib\site-packages
(from requests->torchvision->
treadscan) (3.3)
Requirement already satisfied: charset-normalizer<3,>=2 in
c:\users\asus\appdata\local\programs\python\python310\lib\site-packages
(from requests->t
orchvision->treadscan) (2.1.1)
Requirement already satisfied: certifi>=2017.4.17 in
c:\users\asus\appdata\local\programs\python\python310\lib\site-packages
(from requests->torchvi
sion->treadscan) (2022.6.15)
Requirement already satisfied: urllib3<1.27,>=1.21.1 in
c:\users\asus\appdata\local\programs\python\python310\lib\site-packages
(from requests->torc
hvision->treadscan) (1.26.12)
Requirement already satisfied: six>=1.5 in
c:\users\asus\appdata\local\programs\python\python310\lib\site-packages
(from python-dateutil>=2.7->matpl
otlib>=3.4.2->improutils->treadscan) (1.16.0)
Building wheels for collected packages: treadscan
  Building wheel for treadscan (pyproject.toml) ... error
  error: subprocess-exited-with-error

  × Building wheel for treadscan (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [14 lines of output]
      running bdist_wheel
      running build
      creating build\lib.win-amd64-cpython-310
      creating build\lib.win-amd64-cpython-310\treadscan
      copying src\treadscan\detector.py ->
build\lib.win-amd64-cpython-310\treadscan
      copying src\treadscan\extractor.py ->
build\lib.win-amd64-cpython-310\treadscan
      copying src\treadscan\segmentor.py ->
build\lib.win-amd64-cpython-310\treadscan
      copying src\treadscan\utilities.py ->
build\lib.win-amd64-cpython-310\treadscan
      copying src\treadscan\__init__.py ->
build\lib.win-amd64-cpython-310\treadscan
      running build_ext
      building 'treadscan.cwrap' extension
      error: Microsoft Visual C++ 14.0 or greater is required. Get it with
"Microsoft C++ Build Tools": https://visualstudio.microsoft.com/visual-cp
p-build-tools/
      [end of output]

  note: This error originates from a subprocess, and is likely not a
problem with pip.
  ERROR: Failed building wheel for treadscan
Failed to build treadscan
ERROR: Could not build wheels for treadscan, which is required to install
pyproject.toml-based projects

**Comment by bohundan** [comment]
Yeah, you're missing the Microsoft Visual C++ compiler (the Cython extension must be compiled locally, because compiled binaries aren't transferable between all systems).

But I'm having a different issue with GCC compiler, it doesn't have `Python.h`. On another system that I tried, installing `python3-devel` only installed stuff in `/usr/include/python3.6m` oddly enough. I then cloned that folder to `/usr/include/python3.8/` (the Python version I wanted to install the package on), solving the issue.

This of course shouldn't be required from a user trying to install a python package. Pip and Cython should be taking care of necessary files themselves, as they might also be different for each system and each Python version. And I'm lost as to what, as a package maintainer, am I supposed to do that the Cython extension compiles without problems.

EDIT: Even https://pypi.org/project/cython-package-example/ throws the same error (`Python.h` missing). I guess it is a system-specific issue after all, not a package-specific issue.

Well, now I know how to solve the issue on Linux systems and I might make a guide of some sorts, I have no idea how it works on Windows, sorry (if it even is an issue there).

**Comment by bohundan** [comment]
I've tested installation on another two systems, one Linux, one Windows. In both of them `pip install treadscan` works and the extension compiles without issues. Both systems had Python3.10.

I've also created a short wiki page https://github.com/bohundan/treadscan/wiki summarizing solutions to known installation issues.

As this isn't specific to the `treadscan` package, but to the configuration of the user's system, I'm closing this issue. If you have further issues with installing this (or other packages), try asking a question on stackoverflow, because the users there will be able to assist you more than I can.
